### PR TITLE
repair connection timeout 

### PR DESCRIPTION
**Pull request by d-w-moore**
The session's `connection_timeout` parameter wasn't robustly enough kept or enforced.  As an example, after a `cleanup( )`, the session connection pool, on next use, would return to using the default timeout.

### CODE REVIEW

**Comment by d-w-moore** [comment]
ready for review

File: irods/test/connection_test.py [hunk 1]
```diff
@@ -65,6 +67,47 @@ def test_reply_failure(self):
...
+        sess = self.sess
+
+        # skip if server is remote
+        if sess.host not in ('localhost', socket.gethostname()):
+            self.skipTest('Requires access to server-side file(s)')
+        r = sess.resources.create('hello', 'unixfilesystem', host='localhost')
+        rand = os.urandom(1024)*64
+
+        # Create a large file.
+        size = 1024**3*4
```
Comment (trel) [inline irods/test/connection_test.py:80]: consider adding '4GB' to the comment

File: irods/test/connection_test.py [hunk 2]
```diff
@@ -65,6 +67,47 @@ def test_reply_failure(self):
...
+        physical_path = '/tmp/testfile'
```
Comment (trel) [inline irods/test/connection_test.py:81]: is this windows-friendly?

**Review by trel — COMMENTED** [review]
pretty good - anyone else you'd want to look at this?   in the community?

File: irods/test/connection_test.py [hunk 3]
```diff
@@ -65,6 +67,47 @@ def test_reply_failure(self):
...
+            # Set the connection pool's socket timeout interval back to default, and clean up.
+            sess.connection_timeout = None
+            sess.cleanup()
+            coll.remove(recurse = True, force = True)
+            r.remove()
+            if os.path.isfile(physical_path):
+                os.unlink(physical_path)
+
+    def test_connection_timeout_parameter_in_session_init__issue_377(self):
+            timeout_value = 1.0
```
Comment (alanking) [inline irods/test/connection_test.py:108]: indentation--

File: irods/test/connection_test.py [hunk 4]
```diff
@@ -65,6 +67,47 @@ def test_reply_failure(self):
...
+            s = helpers.make_session(connection_timeout = timeout_value)
+            self.assertEqual(s.pool.connection_timeout, timeout_value)
```
Comment (alanking) [inline irods/test/connection_test.py:110]: Maybe worth adding a section (or a separate test) which tests the `None` case?

**Comment by trel** [comment]
please say when it’s been seen to pass on windows (at least the new test)

**Comment by d-w-moore** [comment]
> please say when it’s been seen to pass on windows (at least the new test)

I actually did see it pass on Windows 10 this afternoon, with the client connected on my home network to an iRODS 4.3.0 server.

File: irods/test/connection_test.py (hunk 4 above)
Comment (d-w-moore) [reply irods/test/connection_test.py:105]: > Maybe worth adding a section (or a separate test) which tests the `None` case?

Possibly, yes.  I have some doubt about the present treatment of defaults, so that might change.   I have now come to the belief that
   - session constructor should default the connection_timeout parameter to the value of `irods.DEFAULT_CONNECTION_TIMEOUT = 120`
   - `None`, like any other value given as the value of this parameter, should filter down as the pool value of `connection_timeout`, and in the case of `None`, it will mean that the socket behaves according to the OS default (as determined by `socket.getdefaulttimeout( )`

**Comment by trel** [comment]
looking pretty good.   please resolve any conversations once 'done'.

File: irods/test/connection_test.py (hunk 2 above)
Comment (d-w-moore) [reply irods/test/connection_test.py]: I have changed to using `NamedTemporaryFile` , and a PUT instead of a REG in the test. so I think we are Windows friendly now.

**Comment by d-w-moore** [comment]
This seems good to go but will test thoroughly later with #447

**Comment by trel** [comment]
please resolve and squash and no pound.

File: irods/test/connection_test.py (hunk 4 above)
Comment (d-w-moore) [reply irods/test/connection_test.py]: Done, at least to the extent that we know the timeout value requested does propagate down to the socket object, and it is verified in a test.

**Comment by d-w-moore** [comment]
> pretty good - anyone else you'd want to look at this? in the community?

I think I'm good, given the passing results in the long-running checksum test.   Originally this was brought to our attention by @jpmcfarland , but his needs seem to have been met, I think.... If not, then I'd hope for his or other community reports and then release a v1.1.9 fairly soon  ; )

Will squash and no #

**Comment by trel** [comment]
let's find another word than 'dysfunctional'.   or simply remove it.

but yes, looks good otherwise.

**Comment by trel** [comment]
i think it's ready - let's # it.
